{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gen559 argparse practice notebook\n",
    "### 2020.12.02"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Practice problem 1 \n",
    "\n",
    "Create a command line script, *proccess-vcf.py* that contains an argument parser. Add the following features to your argument parser:\n",
    "\n",
    "1. Specify the .vcf file to operate on. **This argument should be required.**      \n",
    "\n",
    "   \n",
    "   \n",
    "2. Specifiy the chromosome to extract information about. **This should be required but have a default value of 'chr1'.**  \n",
    "\n",
    "\n",
    "3. Specify which variant type of the possiblities defined below should be extracted. **The user should only be allowed to specify one type, i.e. these should be mutually exclusive options:**\n",
    "   * Single nucleotide change from Reference\n",
    "   * Insertion relative to Reference\n",
    "   * Deletion relative to Reference     \n",
    "   \n",
    "   \n",
    "4. Specify the stem of the name of the output file that will be produced. E.g \"my_variants\" for the output file *my_variants.vcf*\n",
    "\n",
    "Run your script using the Terminal (OS X, Linux, UNIX) or Command Prompt (Windows) on *gm12878.hg38.vcf* and demonstrate its functionality.\n",
    "\n",
    "You may use the cell below to write out your code if you do not feel comfortable using an IDE, but be advised notebooks do not have the ability to properly execute argparse. *Hint: if needed, you can write all of the necessary logic to build your parser as functions in the notebook to test their function.*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Solution\n",
    "\n",
    "* *process-vcf.py* code (shown below) and outputs from running on *gm12878.hg38.vcf* with -c = 'chr17' in '-s', '-i' and '-d' mode are available on [GitHub](https://github.com/beliveau-lab/gen559/tree/main/notebooks) and [Canvas](https://canvas.uw.edu/courses/1430304/files/folder/jupyter_notebooks).\n",
    "\n",
    "\n",
    "**process-vcf.py:**  \n",
    "\n",
    "```python\n",
    "import argparse\n",
    "\n",
    "\n",
    "def extract_variation(file, chrom):\n",
    "    '''Takes in name of vcf file and returns a list of variants\n",
    "    from a specified chromosome'''\n",
    "\n",
    "    # Open specified file.\n",
    "    with open(file, 'r') as f:\n",
    "\n",
    "        # Extract SNVs if args.snv is True.\n",
    "        if args.snv is True:\n",
    "            # Create and populate of extracted variants. Skip header lines.\n",
    "            # Only consider line if length of 'REF' and 'ALT' both = 1, i.e.\n",
    "            # the line describes a SNV.\n",
    "            variants = [line.strip() for line in f \\\n",
    "            if line.strip().split(\"\\t\")[0][0]!='#' \\\n",
    "            and line.strip().split(\"\\t\")[0] == chrom \\\n",
    "            and len(line.strip().split(\"\\t\")[3]) == 1 \\\n",
    "            and len(line.strip().split(\"\\t\")[4]) == 1]\n",
    "\n",
    "        # Extract insertions if args.ins is True.\n",
    "        if args.insertion is True:\n",
    "\n",
    "            # Create and populate of extracted variants. Skip header lines.\n",
    "            # Only consider line if length of 'REF' = 1 and length of 'ALT'\n",
    "            # > 1, i.e. the line describes an insertion.\n",
    "            variants = [line.strip() for line in f \\\n",
    "            if line.strip().split(\"\\t\")[0][0]!='#' \\\n",
    "            and line.strip().split(\"\\t\")[0] == chrom \\\n",
    "            and len(line.strip().split(\"\\t\")[3]) == 1 \\\n",
    "            and len(line.strip().split(\"\\t\")[4]) > 1]\n",
    "\n",
    "        # Extract deletions if args.del is True.\n",
    "        if args.deletion is True:\n",
    "\n",
    "            # Create and populate of extracted variants. Skip header lines.\n",
    "            # Only consider line if length of 'REF' > 1 and length of 'ALT'\n",
    "            # = 1, i.e. the line describes a deletion.\n",
    "            variants = [line.strip() for line in f \\\n",
    "            if line.strip().split(\"\\t\")[0][0]!='#' \\\n",
    "            and line.strip().split(\"\\t\")[0] == chrom \\\n",
    "            and len(line.strip().split(\"\\t\")[3]) > 1 \\\n",
    "            and len(line.strip().split(\"\\t\")[4]) == 1]\n",
    "\n",
    "    # Return list of extracted variants.\n",
    "    return variants\n",
    "\n",
    "\n",
    "# Create argument parser.\n",
    "userInput = argparse.ArgumentParser(description= \\\n",
    "                                    'Takes an input vcf file and extracts ' \\\n",
    "                                    'extracts chromosome-specific variant ' \\\n",
    "                                    'information to return in a new .vcf file')\n",
    "\n",
    "## Add arguments to parser.\n",
    "\n",
    "# Add argument to import file.\n",
    "userInput.add_argument('-f', '--file', action='store', required=True, \\\n",
    "                        help='The .vcf file to process (required)')\n",
    "\n",
    "# Add argument to specify chromosome.\n",
    "userInput.add_argument('-c', '--chrom', action='store', required=True, \\\n",
    "                        default='chr1', help='The chromosome from which ' \\\n",
    "                        'to extract variant information (required), ' \\\n",
    "                        'default=\"chr1\"')\n",
    "\n",
    "# Create mutually exclusive argument group.\n",
    "mutEx = userInput.add_mutually_exclusive_group()\n",
    "\n",
    "# Add SNV argument to mutEx group.\n",
    "mutEx.add_argument('-s', '--snv', action='store_true', default=False,\n",
    "                    help='Return SNV variants, default=False')\n",
    "\n",
    "# Add insertion argument to mutEx group.\n",
    "mutEx.add_argument('-i', '--insertion', action='store_true', default=False, \\\n",
    "                    help='Return insertion variants, default=False')\n",
    "\n",
    "# Add deletion argument to mutEx group.\n",
    "mutEx.add_argument('-d', '--deletion', action='store_true', default=False, \\\n",
    "                    help='Return deletion variants, default=False')\n",
    "\n",
    "# Add argument to specify output filename stem.\n",
    "userInput.add_argument('-o', '--output', action='store', required=True, \\\n",
    "                        default='my_variants', help='The stem of the ' \\\n",
    "                        'output file name (required), ' \\\n",
    "                        'default=\"my_variants\"')\n",
    "\n",
    "# Import user-specified command line values.\n",
    "args = userInput.parse_args()\n",
    "\n",
    "# Call extract_variation function with user input.\n",
    "outputList = extract_variation(args.file, args.chrom)\n",
    "\n",
    "# Create and write output file.\n",
    "with open(args.output + '.vcf', 'w') as f:\n",
    "    f.write('\\n'.join(outputList))\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Practice problem 2\n",
    "\n",
    "Use *process-vcf.py* on *gm12878.hg38.vcf* to extract both insertion and deletion variants from chr17. In the cell below, write code to parse the output files and determine the number and average length of the both the insertion and deletion variants. Print your results. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 8830 insertion variants with an average length of 4.21 bp\n",
      "There are 8531 deletion variants with an average length of 4.06 bp\n"
     ]
    }
   ],
   "source": [
    "# Open insertions file and extact insertion lengths.\n",
    "with open(\"chr17_ins.vcf\", \"r\") as f:\n",
    "    insertion_lengths = [len(line.strip().split(\"\\t\")[4]) for line in f]\n",
    "\n",
    "# Open deletions file and extact deletion lengths.\n",
    "with open(\"chr17_del.vcf\", \"r\") as f:\n",
    "    deletion_lengths = [len(line.strip().split(\"\\t\")[3]) for line in f]\n",
    "\n",
    "\n",
    "# Print request info about variants.\n",
    "print('There are %d insertion variants with an average length of %0.2f bp' \\\n",
    "      % (len(insertion_lengths), sum(insertion_lengths)/len(insertion_lengths)))\n",
    "\n",
    "print('There are %d deletion variants with an average length of %0.2f bp' \\\n",
    "      % (len(deletion_lengths), sum(deletion_lengths)/len(deletion_lengths)))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
